flink-runtime/src/main/java/org/apache/flink/runtime/executiongraph/Execution.java

```diff
@@ -818,71 +800,46 @@ else if (numConsumers == 0) {
 						consumerVertex.checkInputDependencyConstraints()) {
 					scheduleConsumer(consumerVertex);
 				}
-
-				// double check to resolve race conditions
-				if (consumerVertex.getExecutionState() == RUNNING) {
-					consumerVertex.sendPartitionInfos();
-				}
 			}
 			// ----------------------------------------------------------------
 			// Consumer is running => send update message now
+			// Consumer is deploying => cache the partition info which would be
+			// sent after switching to running
 			// ----------------------------------------------------------------
-			else {
-				if (consumerState == RUNNING) {
-					final LogicalSlot consumerSlot = consumer.getAssignedResource();
-
-					if (consumerSlot == null) {
-						// The consumer has been reset concurrently
-						continue;
-					}
-
-					final TaskManagerLocation partitionTaskManagerLocation = partition.getProducer()
-							.getCurrentAssignedResource().getTaskManagerLocation();
-					final ResourceID partitionTaskManager = partitionTaskManagerLocation.getResourceID();
-
-					final ResourceID consumerTaskManager = consumerSlot.getTaskManagerLocation().getResourceID();
-
-					final ResultPartitionID partitionId = new ResultPartitionID(partition.getPartitionId(), attemptId);
-
-					final ResultPartitionLocation partitionLocation;
+			else if (consumerState == DEPLOYING || consumerState == RUNNING) {
+				final LogicalSlot consumerSlot = consumer.getAssignedResource();
+				if (consumerSlot == null) {
```

Comment:
`consumerSlot == null` concurrent change should not be the case anymore as well.
`ASSIGNED_SLOT_UPDATER` in `Execution.tryAssignResource` is called only in main thread, same as `scheduleOrUpdateConsumers`.

Reply:
Yes, I missed this part. :)